### PR TITLE
Fix potential race condition

### PR DESCRIPTION
**Pull request by ryanolson**
Prior to this PR, there was a possible race condition that could be hit by two execution paths where,

- path 1 calls Server::drop_stream which is a yielding call
- path 2 call Server:m_connections->drop_all_streams();

`Server::drop_stream` locks the server's state, but `Server::m_connections` is not protected by that mutex.

Relates to #247

### CODE REVIEW

File: cpp/mrc/src/internal/control_plane/server.cpp [hunk 1]
```diff
@@ -169,7 +169,7 @@ void Server::do_service_await_join()
 {
     // clear all instances which drops their held stream writers
     DVLOG(10) << "awaiting all streams";
-    m_connections.drop_all_streams();
```
Comment (ryanolson) [inline cpp/mrc/src/internal/control_plane/server.cpp:172]: If this PR solves this part of the problem, we should remove `drop_all_streams` from the connection manangers interface.

**Comment by ryanolson** [comment]
Note, this will pass CI, because CI is still skipping test_mrc_private.x.

**Comment by ryanolson** [comment]
We could also convert the Server's

```
mutable boost::fibers::mutex m_mutex;
```

to a 

```
mutable boost::fibers::recursive_mutex m_mutex;

**Review by mdemoret-nv — APPROVED** [review]

**Comment by ryanolson** [comment]
@gpucibot merge

**Pull request merged**